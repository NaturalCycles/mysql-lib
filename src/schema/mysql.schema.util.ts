import { CommonSchema, CommonSchemaField, DATA_TYPE } from '@naturalcycles/db-lib'
import * as mysql from 'mysql'

export interface MySQLSchemaOptions {
  /**
   * @default 'InnoDB'
   */
  engine?: string
}

/**
 * It currently skips nullability and declares everything as "DEFAULT NULL".
 */
export function commonSchemaToMySQLDDL(schema: CommonSchema, opt: MySQLSchemaOptions = {}): string {
  const { engine = 'InnoDB' } = opt

  const { table, fields } = schema

  const lines: string[] = [`CREATE TABLE ${mysql.escapeId(table)} (`]

  const innerLines = fields.map(f => commonSchemaFieldToDDL(f))
  innerLines.push(`PRIMARY KEY (id)`)

  lines.push(innerLines.join(',\n'))
  lines.push(`) ENGINE=${engine}`)

  return lines.join('\n')
}

const typeToMySQLType: Record<DATA_TYPE, string> = {
  [DATA_TYPE.STRING]: `LONGTEXT`,
  [DATA_TYPE.LOCAL_DATE]: `VARCHAR(255)`,
  [DATA_TYPE.INT]: `INT(11)`,
  [DATA_TYPE.TIMESTAMP]: `INT(11)`,
  [DATA_TYPE.FLOAT]: `FLOAT(11)`,
  [DATA_TYPE.BOOLEAN]: `TINYINT(1)`,
  [DATA_TYPE.BINARY]: `BLOB`,
  [DATA_TYPE.ARRAY]: `LONGTEXT`, // will be JSON.stringified
  [DATA_TYPE.OBJECT]: `LONGTEXT`, // will be JSON.stringified
  [DATA_TYPE.NULL]: `VARCHAR(255)`,
  [DATA_TYPE.UNKNOWN]: `LONGTEXT`,
}

function commonSchemaFieldToDDL(f: CommonSchemaField): string {
  if (f.name === 'id') {
    return `id VARCHAR(255) NOT NULL`
  }

  const tokens: string[] = [
    mysql.escapeId(f.name),
    typeToMySQLType[f.type] || typeToMySQLType[DATA_TYPE.UNKNOWN],
    `DEFAULT NULL`,
  ]

  return tokens.join(' ')
}

// CREATE TABLE test_table (
//   id varchar(255) NOT NULL,
//   created int(11) DEFAULT NULL,
//   updated int(11) DEFAULT NULL,
//   _ver int(11) DEFAULT NULL,
//   k1 varchar(255) DEFAULT NULL,
//   k2 varchar(255) DEFAULT NULL,
//   k3 int(11) DEFAULT NULL,
//   even boolean default null,
//   PRIMARY KEY (id)
// ) ENGINE=InnoDB`)
