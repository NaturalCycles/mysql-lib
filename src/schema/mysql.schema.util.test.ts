import { createTestItemsBM, InMemoryDB, TEST_TABLE } from '@naturalcycles/db-lib'
import { commonSchemaToMySQLDDL } from './mysql.schema.util'

test('commonSchemaToMySQLDDL', async () => {
  const items = createTestItemsBM(5)

  const db = new InMemoryDB()
  await db.saveBatch(TEST_TABLE, items)
  const schema = await db.getTableSchema(TEST_TABLE)
  // console.log(schema)

  const ddl = commonSchemaToMySQLDDL(schema)
  // console.log(ddl)
  expect(ddl).toMatchSnapshot()
})
