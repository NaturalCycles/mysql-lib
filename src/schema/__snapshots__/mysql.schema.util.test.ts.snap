// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commonSchemaToMySQLDDL 1`] = `
"CREATE TABLE \`TEST_TABLE\` (
id VARCHAR(255) NOT NULL,
\`k1\` LONGTEXT DEFAULT NULL,
\`k2\` LONGTEXT DEFAULT NULL,
\`k3\` INT(11) DEFAULT NULL,
\`even\` TINYINT(1) DEFAULT NULL,
\`created\` INT(11) DEFAULT NULL,
\`updated\` INT(11) DEFAULT NULL,
PRIMARY KEY (id)
) ENGINE=InnoDB"
`;
